`TestBackendScore` failed

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftiproxy%2Fghpr_unit_test/detail/ghpr_unit_test/40/pipeline

### 2. What did you expect to see? (Required)

No error

### 3. What did you see instead (Required)

```
--- FAIL: TestBackendScore (0.00s)

    logger.go:25: 2024-01-23T17:29:13.072+0800	INFO	TestBackendScore	update backend	{"backend_addr": "1", "prev": "none", "cur": "status: healthy"}

    logger.go:25: 2024-01-23T17:29:13.072+0800	INFO	TestBackendScore	update backend	{"backend_addr": "2", "prev": "none", "cur": "status: healthy"}

    logger.go:25: 2024-01-23T17:29:13.072+0800	INFO	TestBackendScore	update backend	{"backend_addr": "3", "prev": "none", "cur": "status: healthy"}

    logger.go:25: 2024-01-23T17:29:13.072+0800	INFO	TestBackendScore	update backend	{"backend_addr": "1", "prev": "status: healthy", "cur": "status: down"}

    logger.go:25: 2024-01-23T17:29:13.072+0800	INFO	TestBackendScore	update backend	{"backend_addr": "2", "prev": "status: healthy", "cur": "status: down"}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	update backend	{"backend_addr": "5", "prev": "none", "cur": "status: healthy"}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	update backend	{"backend_addr": "6", "prev": "none", "cur": "status: healthy"}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	update backend	{"backend_addr": "4", "prev": "none", "cur": "status: healthy"}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 1, "from": "3", "to": "4", "from_score": 90, "to_score": 0}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 2, "from": "3", "to": "6", "from_score": 89, "to_score": 0}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 4, "from": "3", "to": "5", "from_score": 88, "to_score": 0}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 5, "from": "3", "to": "5", "from_score": 87, "to_score": 1}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 7, "from": "3", "to": "6", "from_score": 86, "to_score": 1}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 9, "from": "3", "to": "4", "from_score": 85, "to_score": 1}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 10, "from": "3", "to": "4", "from_score": 84, "to_score": 2}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 12, "from": "3", "to": "6", "from_score": 83, "to_score": 2}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 13, "from": "3", "to": "5", "from_score": 82, "to_score": 2}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 14, "from": "3", "to": "5", "from_score": 81, "to_score": 3}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 15, "from": "3", "to": "6", "from_score": 80, "to_score": 3}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 16, "from": "3", "to": "4", "from_score": 79, "to_score": 3}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 17, "from": "3", "to": "4", "from_score": 78, "to_score": 4}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 18, "from": "3", "to": "6", "from_score": 77, "to_score": 4}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 19, "from": "3", "to": "5", "from_score": 76, "to_score": 4}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 20, "from": "3", "to": "5", "from_score": 75, "to_score": 5}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 21, "from": "3", "to": "6", "from_score": 74, "to_score": 5}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 22, "from": "3", "to": "4", "from_score": 73, "to_score": 5}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 23, "from": "3", "to": "4", "from_score": 72, "to_score": 6}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 24, "from": "3", "to": "6", "from_score": 71, "to_score": 6}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 25, "from": "3", "to": "5", "from_score": 70, "to_score": 6}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 26, "from": "3", "to": "5", "from_score": 69, "to_score": 7}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 27, "from": "3", "to": "6", "from_score": 68, "to_score": 7}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 28, "from": "3", "to": "4", "from_score": 67, "to_score": 7}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 29, "from": "3", "to": "4", "from_score": 66, "to_score": 8}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 30, "from": "3", "to": "6", "from_score": 65, "to_score": 8}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 31, "from": "3", "to": "5", "from_score": 64, "to_score": 8}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 32, "from": "3", "to": "5", "from_score": 63, "to_score": 9}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 33, "from": "3", "to": "6", "from_score": 62, "to_score": 9}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 34, "from": "3", "to": "4", "from_score": 61, "to_score": 9}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 35, "from": "3", "to": "4", "from_score": 60, "to_score": 10}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 36, "from": "3", "to": "6", "from_score": 59, "to_score": 10}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 37, "from": "3", "to": "5", "from_score": 58, "to_score": 10}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 38, "from": "3", "to": "5", "from_score": 57, "to_score": 11}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 39, "from": "3", "to": "6", "from_score": 56, "to_score": 11}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 40, "from": "3", "to": "4", "from_score": 55, "to_score": 11}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 41, "from": "3", "to": "4", "from_score": 54, "to_score": 12}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 42, "from": "3", "to": "6", "from_score": 53, "to_score": 12}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 43, "from": "3", "to": "5", "from_score": 52, "to_score": 12}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 44, "from": "3", "to": "5", "from_score": 51, "to_score": 13}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 46, "from": "3", "to": "6", "from_score": 50, "to_score": 13}

    logger.go:25: 2024-01-23T17:29:13.073+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 47, "from": "3", "to": "4", "from_score": 49, "to_score": 13}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 48, "from": "3", "to": "4", "from_score": 48, "to_score": 14}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 49, "from": "3", "to": "6", "from_score": 47, "to_score": 14}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 50, "from": "3", "to": "5", "from_score": 46, "to_score": 14}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 52, "from": "3", "to": "5", "from_score": 45, "to_score": 15}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 53, "from": "3", "to": "6", "from_score": 44, "to_score": 15}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 54, "from": "3", "to": "4", "from_score": 43, "to_score": 15}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 55, "from": "3", "to": "4", "from_score": 42, "to_score": 16}

    logger.go:25: 2024-01-23T17:29:13.074+0800	INFO	TestBackendScore	begin redirect connection	{"connID": 56, "from": "3", "to": "6", "from_score": 41, "to_score": 16}

    router_test.go:199: 

        	Error Trace:	/home/jenkins/agent/workspace/pingcap/tiproxy/ghpr_unit_test/tiproxy/pkg/manager/router/router_test.go:199

        	            				/home/jenkins/agent/workspace/pingcap/tiproxy/ghpr_unit_test/tiproxy/pkg/manager/router/router_test.go:247

        	            				/home/jenkins/agent/workspace/pingcap/tiproxy/ghpr_unit_test/tiproxy/pkg/manager/router/router_test.go:348

        	Error:      	"13" is not greater than or equal to "16"

        	Test:       	TestBackendScore

FAIL
```

### 4. What is your version? (Required)

<!-- Paste the output of tiproxy version -->
<!-- Paste the output of tiproxyctl version if related -->
<!-- Paste the output of SELECT tidb_version()  if related -->

master

--- a/pkg/manager/router/router_score.go
+++ b/pkg/manager/router/router_score.go
@@ -279,6 +279,9 @@ func (router *ScoreBasedRouter) OnConnClosed(addr string, conn RedirectableConn)
 	if redirectingBackend != nil {
 		redirectingBackend.connScore--
 		connWrapper.Value.redirectingBackend = nil
+		if redirectingBe := router.lookupBackend(redirectingBackend.addr, true); redirectingBe != nil {
+			router.adjustBackendList(redirectingBe, true)
+		}
 	} else {
 		be.Value.connScore--
 	}

